Spigot-Server-Patches/0721-Have-CraftMerchantCustom-emit-PlayerPurchaseEvent.patch

```diff
@@ -0,0 +1,83 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Alexander <protonull@protonmail.com>
+Date: Thu, 6 May 2021 13:01:25 +0100
+Subject: [PATCH] Have CraftMerchantCustom emit PlayerPurchaseEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/item/trading/IMerchant.java b/src/main/java/net/minecraft/world/item/trading/IMerchant.java
+index 5644affeadc51ac3209ee9a087b3d2dd887e401c..7c54ca6af424b9d31f73187df9015f5df84762f4 100644
+--- a/src/main/java/net/minecraft/world/item/trading/IMerchant.java
++++ b/src/main/java/net/minecraft/world/item/trading/IMerchant.java
+@@ -19,7 +19,7 @@ public interface IMerchant {
+ 
+     MerchantRecipeList getOffers();
+ 
+-    void a(MerchantRecipe merchantrecipe);
++    void a(MerchantRecipe merchantrecipe); default void handlePurchase(MerchantRecipe merchantRecipe) { a(merchantRecipe); } // Paper - OBFHELPER
+ 
+     void k(ItemStack itemstack);
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMerchantCustom.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMerchantCustom.java
+index 206c133ebc6c44038585236b0628543b8bed278c..3d05c5e76b860231e863108b3372fe82c672d405 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMerchantCustom.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMerchantCustom.java
+@@ -1,9 +1,15 @@
+ package org.bukkit.craftbukkit.inventory;
+ 
++import io.papermc.paper.event.player.PlayerPurchaseEvent;
+ import net.minecraft.network.chat.ChatComponentText;
+ import net.minecraft.network.chat.IChatBaseComponent;
++import net.minecraft.server.level.EntityPlayer;
+ import net.minecraft.sounds.SoundEffect;
+ import net.minecraft.sounds.SoundEffects;
++import net.minecraft.world.entity.Entity;
++import net.minecraft.world.entity.EntityExperienceOrb;
++import net.minecraft.world.entity.npc.EntityVillagerAbstract;
++import net.minecraft.world.entity.npc.EntityVillagerTrader;
+ import net.minecraft.world.entity.player.EntityHuman;
+ import net.minecraft.world.item.ItemStack;
+ import net.minecraft.world.item.trading.IMerchant;
+@@ -11,6 +17,7 @@ import net.minecraft.world.item.trading.MerchantRecipe;
+ import net.minecraft.world.item.trading.MerchantRecipeList;
+ import net.minecraft.world.level.World;
+ import org.apache.commons.lang.Validate;
++import org.bukkit.entity.ExperienceOrb;
```

Comment:
I thought we were still not supposed to just import things? Pretty sure I saw kashike say that somewhere.

Reply:
What should I do as an alternative?

Reply:
just use fully qualified names in the code. Yeah, its annoying, and maybe core team changed their minds on this, but last I heard, to minimize diff, its best to avoid imports where you can.

Reply:
Done :)